sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/sqs/SqsUnboundedReaderTest.java

```diff
@@ -17,113 +17,155 @@
-    source =
-        new SqsUnboundedSource(
-            SqsIO.read()
-                .withQueueUrl(queueUrl)
-                .withSqsClientProvider(SqsClientProviderMock.of(client))
-                .withMaxNumRecords(1));
-  }
-
-  private void setupMessages(List<String> messages) {
-    final SqsClient client = embeddedSqsRestServer.getClient();
-    final String queueUrl = embeddedSqsRestServer.getQueueUrl();
+  private void setupMessages(String... messages) {
+    final SqsClient client = testCase.getClient();
+    final String queueUrl = testCase.getQueueUrl();
     for (String message : messages) {
-      client.sendMessage(
-          SendMessageRequest.builder().queueUrl(queueUrl).messageBody(message).build());
+      client.sendMessage(b -> b.queueUrl(queueUrl).messageBody(message));
     }
-    source =
-        new SqsUnboundedSource(
-            SqsIO.read()
-                .withQueueUrl(queueUrl)
-                .withSqsClientProvider(SqsClientProviderMock.of(client))
-                .withMaxNumRecords(messages.size()));
+
+    when(mockSource.getRead().sqsClientProvider()).thenReturn(SqsClientProviderMock.of(client));
+    when(mockSource.getRead().queueUrl()).thenReturn(queueUrl);
   }
 
   @Test
   public void testReadOneMessage() throws IOException {
-    setupOneMessage();
-    UnboundedSource.UnboundedReader<SqsMessage> reader =
-        source.createReader(pipeline.getOptions(), null);
+    setupMessages(DATA);
+    SqsUnboundedReader reader = new SqsUnboundedReader(mockSource, null);
+
     // Read one message.
     assertTrue(reader.start());
     assertEquals(DATA, reader.getCurrent().getBody());
     assertFalse(reader.advance());
+
     // ACK the message.
     UnboundedSource.CheckpointMark checkpoint = reader.getCheckpointMark();
     checkpoint.finalizeCheckpoint();
     reader.close();
   }
 
   @Test
-  public void testTimeoutAckAndRereadOneMessage() throws IOException {
-    setupOneMessage();
-    UnboundedSource.UnboundedReader<SqsMessage> reader =
-        source.createReader(pipeline.getOptions(), null);
-    SqsClient sqsClient = embeddedSqsRestServer.getClient();
+  public void testAckDeletedMessage() throws IOException {
+    setupMessages(DATA);
+    SqsUnboundedReader reader = new SqsUnboundedReader(mockSource, null);
+
+    // Read one message
+    assertTrue(reader.start());
+    assertEquals(DATA, reader.getCurrent().getBody());
+    String receiptHandle = reader.getCurrent().getReceiptHandle();
+    assertFalse(reader.advance());
+
+    // Simulate already ACKed message after re-delivery to different reader
+    testCase
+        .getClient()
+        .deleteMessage(b -> b.queueUrl(testCase.getQueueUrl()).receiptHandle(receiptHandle));
+
+    // Now ACK the message.
+    UnboundedSource.CheckpointMark checkpoint = reader.getCheckpointMark();
+    checkpoint.finalizeCheckpoint();
+    reader.close();
+  }
+
+  @Test
+  @Ignore("Behavior of SQSRestServer is broken")
```

Comment:
why is this broken ? why adding a test that is already broken ? What lacks for fixing ?

Reply:
The batch API of elasticmq used for testing here isn't behaving correctly in this case :/

It's supposed to test the same fix for changing visibility timeouts as tested above for deleting messages.
When I find time I'm intending to fix the issue in elasticmq. For now I've left the test case primarily for documentation.

Reply:
ok, so please create a ticket in elasticmq project refer it in a new Beam ticket about enabling the test back and refer that Beam ticket in the Ignore annotation.

Reply:
done 👍